src/nbl/video/COpenGL_Swapchain.cpp

```diff
@@ -0,0 +1,414 @@
+            syncs[imgix] = core::make_smart_refctd_ptr<COpenGLSync>();
+            syncs[imgix]->init(core::smart_refctd_ptr<IOpenGL_LogicalDevice>(m_device), gl, false);
+            // swap buffers performs an implicit flush before swapping 
+            // https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml
+            egl->call.peglSwapBuffers(egl->display, glctx.surface);
+        }
+    }
+
+    void exit(SThreadHandlerInternalState* state)
+    {
+        // TODO this should be in destructor for COpenGLImage
+        auto gl = getFunctionPointer(m_device->getAPIType(), state);
+        gl->~IOpenGL_FunctionTable();
+
+        egl->call.peglMakeCurrent(egl->display, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+        egl->call.peglDestroyContext(egl->display, glctx.ctx);
+        egl->call.peglDestroySurface(egl->display, glctx.surface);
+    }
+
+    bool wakeupPredicate() const { return needToBlit; }
+    bool continuePredicate() const { return needToBlit; }
+
+private:
+    core::smart_refctd_ptr<IOpenGL_LogicalDevice> m_device;
+    uint64_t m_masterContextCallsWaited;
+
+    const egl::CEGL* egl;
+    ISurface::E_PRESENT_MODE m_presentMode;
+    const COpenGLFeatureMap* features;
+    uint32_t imageCount;
+    uint32_t imgWidth, imgHeight;
+    uint32_t fbos[MaxImages]{};
+    core::smart_refctd_ptr<COpenGLSync> syncs[MaxImages];
+    COpenGLDebugCallback* m_dbgCb;
+    std::array<uint32_t, MaxImages> m_texViews;
+
+    bool needToBlit = false;
+
+    EGLBoolean m_makeCurrentRes = EGL_FALSE;
+    std::condition_variable m_ctxCreatedCvar;
+};
+
+template <typename FunctionTableType_>
+COpenGL_Swapchain<FunctionTableType_>::COpenGL_Swapchain<FunctionTableType_>(
+    SCreationParams&& params,
+    core::smart_refctd_ptr<IOpenGL_LogicalDevice>&& dev,
+    const egl::CEGL* _egl,
+    uint32_t imgCount, IGPUImage::SCreationParams imgCreationParams,
+    const COpenGLFeatureMap* _features,
+    EGLContext _ctx,
+    EGLConfig _config,
+    COpenGLDebugCallback* _dbgCb
+) : ISwapchain(core::smart_refctd_ptr<const ILogicalDevice>(dev), std::move(params), imgCount)
+{
+    _egl->call.peglBindAPI(dev->getEGLAPI());
+
+    const EGLint surface_attributes[] = {
+        EGL_RENDER_BUFFER, EGL_BACK_BUFFER,
+        // EGL_GL_COLORSPACE is supported only for EGL 1.5 and later
+        _egl->version.minor >= 5 ? EGL_GL_COLORSPACE : EGL_NONE, EGL_GL_COLORSPACE_SRGB,
+
+        EGL_NONE
+    };
+
+    auto surface = _egl->call.peglCreateWindowSurface(_egl->display, _config, (EGLNativeWindowType)m_params.surface->getNativeWindowHandle(), surface_attributes);
+    assert(surface != EGL_NO_SURFACE);
+
+    m_threadHandler = std::unique_ptr<COpenGL_SwapchainThreadHandler>(new COpenGL_SwapchainThreadHandler(
+        _egl, dev, m_params.surface->getNativeWindowHandle(), m_params.presentMode, imgCount, imgCreationParams.extent.width, imgCreationParams.extent.height, _features, _ctx, surface, _config, _dbgCb
+    ));
```

Comment:
hoist out of constructor into the `create` call, as this can all fail, and shouldn't assert just because it didnt work